zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -0,0 +1,3865 @@
+    q_blk_hdr_rlc_start: Selector,
+    q_blk_hdr_rlp_end: Selector,
+    blk_hdr_rlc_acc: Column<Advice>,
+    q_blk_hdr_rlc_acc: Column<Advice>,
+    blk_hdr_is_leading_zero: Column<Advice>,
+}
+
+/// PublicData contains all the values that the PiCircuit recieves as input
+#[derive(Debug, Clone, Default)]
+pub struct PublicData<F: Field> {
+    /// chain id
+    pub chain_id: Word,
+    /// History hashes contains the most recent 256 block hashes in history,
+    /// where the latest one is at history_hashes[history_hashes.len() - 1].
+    pub history_hashes: Vec<Word>,
+    /// Block Transactions
+    pub transactions: Vec<eth_types::Transaction>,
+    /// Block State Root
+    pub state_root: H256,
+    /// Previous block root
+    pub prev_state_root: H256,
+    /// Constants related to Ethereum block
+    pub block_constants: BlockConstants,
+
+    /// Prover address
+    pub prover: Address,
+
+    /// Parent hash
+    pub parent_hash: H256,
+    /// The author
+    pub beneficiary: Address,
+    /// Transactions Root
+    pub transactions_root: H256,
+    /// Receipts Root
+    pub receipts_root: H256,
+    /// Logs Bloom
+    // pub logs_bloom: Bloom,
+    /// Gas Used
+    pub gas_used: U256,
+    /// Mix Hash
+    pub mix_hash: H256,
+    /// Withdrawals Root
+    pub withdrawals_root: H256,
+
+    // private values
+    block_rlp: Bytes,
+    block_hash: H256,
+    block_hash_hi: F,
+    block_hash_lo: F,
+
+    txs_rlp: Bytes,
+    txs_hash: H256,
+    txs_hash_hi: F,
+    txs_hash_lo: F,
+
+    blockhash_blk_hdr_rlp: Bytes,
+    blockhash_rlp_hash_hi: F,
+    blockhash_rlp_hash_lo: F,
+}
+
+pub(super) fn rlp_opt<T: rlp::Encodable>(rlp: &mut rlp::RlpStream, opt: &Option<T>) {
+    if let Some(inner) = opt {
+        rlp.append(inner);
+    } else {
+        rlp.append(&"");
+    }
+}
+
+impl<F: Field> PublicData<F> {
+    fn get_block_rlp_rlc(&self, challenges: &Challenges<Value<F>>) -> Value<F> {
+        use crate::evm_circuit::util::rlc;
+        let randomness = challenges.keccak_input();
+        randomness.map(|randomness| rlc::value(self.block_rlp.iter().rev(), randomness))
+    }
+
+    fn get_txs_rlp_rlc(&self, challenges: &Challenges<Value<F>>) -> Value<F> {
+        use crate::evm_circuit::util::rlc;
+        let randomness = challenges.keccak_input();
+        randomness.map(|randomness| rlc::value(self.txs_rlp.iter().rev(), randomness))
+    }
+
+    fn split_hash(hash: [u8; 32]) -> (F, F) {
+        let hi = hash.iter().take(16).fold(F::zero(), |acc, byte| {
+            acc * F::from(BYTE_POW_BASE) + F::from(*byte as u64)
+        });
+
+        let lo = hash.iter().skip(16).fold(F::zero(), |acc, byte| {
+            acc * F::from(BYTE_POW_BASE) + F::from(*byte as u64)
+        });
+        (hi, lo)
+    }
+
+    fn get_txs_hash(txs_rlp: &Bytes) -> (H256, F, F) {
+        let hash = keccak256(&txs_rlp);
+        let (hi, lo) = Self::split_hash(hash);
+        (hash.into(), hi, lo)
+    }
+
+    fn get_block_hash(
```

Comment:
I added a different version of this function instead of modifying in place as I wasn't sure what the intentions are for this one. Should be removed if duplicate now

Reply:
Need your input on this @johntaiko @smtmfft